File: Python3.x/136-Single Number.py
'''
# 列表
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        tmp = []
        for num in nums:
            if num in tmp: tmp.remove(num)
            else: tmp.append(num)
        return tmp[0]
'''


# math
class Solution:
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return 2 * sum(set(nums)) - sum(nums)


'''
2 * (a + b + c) - (a + a + b + b +c) = c  # 其中c是奇数个，a和b是偶数个
'''

File: Algorithms/C++/136-Single_Number.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        map<int, int> m;
        for (int num : nums) {
            if (m.count(num))
                m[num]++;
            else
                m[num] = 1;
        }
        map<int, int>::iterator it;
        int res = 0;
        for (it = m.begin(); it != m.end(); ++it) {
            if ((*it).second == 1)
                res = (*it).first;
        }
        return res;
    }
};


class Solution {
public:
    int singleNumber(vector<int>& nums) {
        vector<int> tmp;
        for (int num : nums) {
            bool flag = true;
            for (vector<int>::iterator it = tmp.begin(); it != tmp.end(); ++it) {
                if (*it == num) {
                    tmp.erase(it);
                    flag = false;
                    break;
                }
            }
            if (flag)
                tmp.push_back(num);
        }
        return tmp[0];
    }
};
